File: 2022/generate.py
#!/usr/bin/env python3

import os
import sys
import shutil


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(f"Usage: {os.path.basename(__file__)} day")
        sys.exit(1)

    day = sys.argv[1]

    formatted_day = day.zfill(2)
    dir = os.path.dirname(os.path.realpath(__file__))

    day_dir = f"{dir}/{formatted_day}"

    if os.path.isdir(day_dir):
        overwrite = input(f"Day {day} already exists. Overwrite? [y/N] ")
        if overwrite.lower() in ["y", "yes"]:
            shutil.rmtree(day_dir)
        else:
            sys.exit(0)

    os.mkdir(day_dir)

    with open(f"{day_dir}/__init__.py", "w"):
        pass

    with open(f"{day_dir}/input.txt", "w"):
        pass

    with open(f"{day_dir}/sample.txt", "w"):
        pass

    with open(f"{day_dir}/solutions.py", "w") as f:
        f.write(
            "\n".join(
                [
                    "def part1(data):",
                    "    return 0",
                    "",
                    "",
                    "def part2(data):",
                    "    return 0",
                    "",
                ]
            )
        )

    with open(f"{day_dir}/test.py", "w") as f:
        f.write(
            "\n".join(
                [
                    "import os",
                    "import unittest",
                    "from .solutions import part1, part2",
                    "",
                    "dir = os.path.dirname(os.path.realpath(__file__))",
                    'with open(f"{dir}/sample.txt", "r") as f:',
                    "    sample_data = f.read()",
                    'with open(f"{dir}/input.txt", "r") as f:',
                    "    input_data = f.read()",
                    "",
                    "",
                    f"class TestDay{formatted_day}Solutions(unittest.TestCase):",
                    "    def test_part1_sample(self):",
                    "        self.assertEqual(part1(sample_data), 0)",
                    "",
                    "    def test_part1_input(self):",
                    "        self.assertEqual(part1(input_data), 0)",
                    "",
                    "    def test_part2_sample(self):",
                    "        self.assertEqual(part2(sample_data), 0)",
                    "",
                    "    def test_part2_input(self):",
                    "        self.assertEqual(part2(input_data), 0)",
                    "",
                    "",
                    'if __name__ == "__main__":',
                    "    unittest.main()",
                    "",
                ]
            )
        )

File: 2023/generate.rb
# frozen_string_literal: true

require 'fileutils'

if ARGV.empty?
  puts "Usage: #{__FILE__} day"
  exit(1)
end

day = format('%02d', ARGV[0])
day_dir = File.join(__dir__, "day#{day}")

if File.directory?(day_dir)
  print "Day #{day.to_i} already exists. Overwrite? [y/N] "
  overwrite = STDIN.gets.chomp

  if %w[y yes].include? overwrite.strip.downcase
    FileUtils.rm_rf(day_dir)
  else
    exit(0)
  end
end

FileUtils.mkdir(day_dir)

FileUtils.touch(File.join(day_dir, 'sample.txt'))

FileUtils.touch(File.join(day_dir, 'input.txt'))

File.open(File.join(day_dir, 'solutions.rb'), 'w') do |f|
  code = <<~RUBY
    # frozen_string_literal: true

    # Solutions for Advent of Code 2023.#{day.to_i}
    module Day#{day}
      module_function

      def part1(input)
        _data = input.each_line(chomp: true).filter { |line| line != '' }
        1
      end

      def part2(input)
        _data = input.each_line(chomp: true).filter { |line| line != '' }
        2
      end
    end
  RUBY

  f.write(code)
end

File.open(File.join(day_dir, 'test.rb'), 'w') do |f|
  code = <<~RUBY
    # frozen_string_literal: true

    require 'minitest/autorun'
    require_relative 'solutions'

    # Tests for Advent of Code 2023.#{day.to_i}
    class TestDay#{day} < Minitest::Test
      def setup
        @sample_data = File.read("\#{__dir__}/sample.txt")
        @input_data = File.read("\#{__dir__}/input.txt")
      end

      def test_part1_sample
        assert_equal(1, Day#{day}.part1(@sample_data))
      end

      def test_part1_input
        assert_equal(1, Day#{day}.part1(@input_data))
      end

      def test_part2_sample
        assert_equal(2, Day#{day}.part2(@sample_data))
      end

      def test_part2_input
        assert_equal(2, Day#{day}.part2(@input_data))
      end
    end
  RUBY

  f.write(code)
end
